File: launcher/ClatterDrive.Launcher/LauncherViewModel.cs
using System;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.IO;
using System.Runtime.CompilerServices;
using System.Windows.Input;

namespace ClatterDrive.Launcher;

public sealed class LauncherViewModel : INotifyPropertyChanged, IDisposable
{
    private readonly IBackendController backend;
    private string backingDirectory = Path.GetFullPath("backing_storage");
    private string host = "127.0.0.1";
    private int port = 8080;
    private string driveProfile = "seagate_ironwolf_pro_16tb";
    private string acousticProfile = "mounted_in_case";
    private string audioMode = "live";
    private string audioDevice = "";
    private string status = "Stopped";
    private bool isRunning;
    private bool coldStart = true;
    private bool asyncPowerOn = true;

    public LauncherViewModel()
        : this(new BackendController())
    {
    }

    public LauncherViewModel(IBackendController backend)
    {
        this.backend = backend;
        ApplyEnvironmentDefaults();
        this.backend.LogReceived += (_, line) => AddLog(line);
        this.backend.Ready += (_, _) => Status = "Running";
        this.backend.Exited += (_, _) =>
        {
            IsRunning = false;
            Status = "Stopped";
        };
        StartCommand = new RelayCommand(_ => Start(), _ => CanStart);
        StopCommand = new RelayCommand(_ => Stop(), _ => IsRunning);
    }

    public event PropertyChangedEventHandler? PropertyChanged;

    public string[] DriveProfiles => ProfileCatalog.DriveProfiles;
    public string[] AcousticProfiles => ProfileCatalog.AcousticProfiles;
    public string[] AudioModes => ProfileCatalog.AudioModes;
    public ObservableCollection<string> Logs { get; } = [];
    public ICommand StartCommand { get; }
    public ICommand StopCommand { get; }

    public string BackingDirectory
    {
        get => backingDirectory;
        set
        {
            if (SetField(ref backingDirectory, value))
            {
                RefreshValidation();
            }
        }
    }

    public string Host
    {
        get => host;
        set
        {
            if (SetField(ref host, value))
            {
                OnPropertyChanged(nameof(WebDavUrl));
                OnPropertyChanged(nameof(NetUseCommand));
                RefreshValidation();
            }
        }
    }

    public int Port
    {
        get => port;
        set
        {
            if (SetField(ref port, value))
            {
                OnPropertyChanged(nameof(WebDavUrl));
                OnPropertyChanged(nameof(NetUseCommand));
                RefreshValidation();
            }
        }
    }

    public string DriveProfile
    {
        get => driveProfile;
        set => SetField(ref driveProfile, value);
    }

    public string AcousticProfile
    {
        get => acousticProfile;
        set => SetField(ref acousticProfile, value);
    }

    public string AudioMode
    {
        get => audioMode;
        set => SetField(ref audioMode, value);
    }

    public string AudioDevice
    {
        get => audioDevice;
        set => SetField(ref audioDevice, value);
    }

    public string Status
    {
        get => status;
        private set => SetField(ref status, value);
    }

    public bool IsRunning
    {
        get => isRunning;
        private set
        {
            if (SetField(ref isRunning, value))
            {
                OnPropertyChanged(nameof(CanStart));
                ((RelayCommand)StartCommand).RaiseCanExecuteChanged();
                ((RelayCommand)StopCommand).RaiseCanExecuteChanged();
            }
        }
    }

    public bool CanStart => !IsRunning && string.IsNullOrEmpty(ValidationMessage);
    public string ValidationMessage => Validate();
    public string WebDavUrl => CurrentSettings().WebDavUrl;
    public string NetUseCommand => MountCommandBuilder.NetUseCommand(CurrentSettings());
    public string NetUseUnmountCommand => MountCommandBuilder.NetUseUnmountCommand();

    public BackendSettings CurrentSettings()
    {
        return new BackendSettings
        {
            Host = Host,
            Port = Port,
            BackingDirectory = BackingDirectory,
            AudioMode = AudioMode,
            AudioDevice = string.IsNullOrWhiteSpace(AudioDevice) ? null : AudioDevice,
            DriveProfile = DriveProfile,
            AcousticProfile = AcousticProfile,
            ColdStart = coldStart,
            AsyncPowerOn = asyncPowerOn,
        };
    }

    public void Start()
    {
        var validation = ValidationMessage;
        if (!string.IsNullOrEmpty(validation))
        {
            Status = validation;
            return;
        }
        Directory.CreateDirectory(BackingDirectory);
        Status = "Starting";
        backend.Start(CurrentSettings());
        IsRunning = true;
    }

    public void Stop()
    {
        backend.Stop();
        IsRunning = false;
        Status = "Stopped";
    }

    public void Dispose()
    {
        backend.Dispose();
    }

    private void AddLog(string line)
    {
        App.Current.Dispatcher.Invoke(() =>
        {
            Logs.Add(line);
            while (Logs.Count > 300)
            {
                Logs.RemoveAt(0);
            }
        });
    }

    private void ApplyEnvironmentDefaults()
    {
        backingDirectory = Environment.GetEnvironmentVariable("CLATTERDRIVE_LAUNCHER_BACKING_DIR") ?? backingDirectory;
        host = Environment.GetEnvironmentVariable("CLATTERDRIVE_LAUNCHER_HOST") ?? host;
        if (int.TryParse(Environment.GetEnvironmentVariable("CLATTERDRIVE_LAUNCHER_PORT"), out var envPort))
        {
            port = envPort;
        }
        audioMode = Environment.GetEnvironmentVariable("CLATTERDRIVE_LAUNCHER_AUDIO") ?? audioMode;
        audioDevice = Environment.GetEnvironmentVariable("CLATTERDRIVE_LAUNCHER_AUDIO_DEVICE") ?? audioDevice;
        driveProfile = Environment.GetEnvironmentVariable("CLATTERDRIVE_LAUNCHER_DRIVE_PROFILE") ?? driveProfile;
        acousticProfile = Environment.GetEnvironmentVariable("CLATTERDRIVE_LAUNCHER_ACOUSTIC_PROFILE") ?? acousticProfile;
        if (IsTruthy(Environment.GetEnvironmentVariable("CLATTERDRIVE_LAUNCHER_READY")))
        {
            coldStart = false;
        }
        if (IsTruthy(Environment.GetEnvironmentVariable("CLATTERDRIVE_LAUNCHER_SYNC_POWER_ON")))
        {
            asyncPowerOn = false;
        }
    }

    private static bool IsTruthy(string? value)
    {
        return value?.Trim().ToLowerInvariant() is "1" or "true" or "yes" or "on";
    }

    private string Validate()
    {
        if (string.IsNullOrWhiteSpace(BackingDirectory))
        {
            return "Choose a backing folder.";
        }
        if (string.IsNullOrWhiteSpace(Host))
        {
            return "Enter a host.";
        }
        if (Port is < 1 or > 65535)
        {
            return "Choose a port from 1 to 65535.";
        }
        return "";
    }

    private void RefreshValidation()
    {
        OnPropertyChanged(nameof(ValidationMessage));
        OnPropertyChanged(nameof(CanStart));
        ((RelayCommand)StartCommand).RaiseCanExecuteChanged();
    }

    private bool SetField<T>(ref T field, T value, [CallerMemberName] string? propertyName = null)
    {
        if (Equals(field, value))
        {
            return false;
        }
        field = value;
        OnPropertyChanged(propertyName);
        return true;
    }

    private void OnPropertyChanged([CallerMemberName] string? propertyName = null)
    {
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    }
}

File: launcher/ClatterDrive.MacLauncher/Sources/ClatterDriveMacCore/LauncherViewModel.swift
import Combine
import Foundation

@MainActor
public final class LauncherViewModel: ObservableObject {
    @Published public var backingDirectory: String
    @Published public var host: String
    @Published public var portText: String
    @Published public var driveProfile: String
    @Published public var acousticProfile: String
    @Published public var audioMode: AudioMode
    @Published public var audioDevice: String
    @Published public var status: String
    @Published public var isRunning: Bool
    @Published public var logs: [String]
    @Published public var driveProfileNames: [String]
    @Published public var acousticProfileNames: [String]

    private let backend: BackendControlling

    public init(backend: BackendControlling = BackendController()) {
        self.backend = backend
        let defaults = BackendSettings()
        backingDirectory = ProcessInfo.processInfo.environment["CLATTERDRIVE_LAUNCHER_BACKING_DIR"] ?? defaults.backingDirectory
        host = ProcessInfo.processInfo.environment["CLATTERDRIVE_LAUNCHER_HOST"] ?? defaults.host
        portText = ProcessInfo.processInfo.environment["CLATTERDRIVE_LAUNCHER_PORT"] ?? String(defaults.port)
        driveProfile = ProcessInfo.processInfo.environment["CLATTERDRIVE_LAUNCHER_DRIVE_PROFILE"] ?? defaults.driveProfile
        acousticProfile = ProcessInfo.processInfo.environment["CLATTERDRIVE_LAUNCHER_ACOUSTIC_PROFILE"] ?? defaults.acousticProfile
        audioMode = AudioMode(rawValue: ProcessInfo.processInfo.environment["CLATTERDRIVE_LAUNCHER_AUDIO"] ?? "") ?? defaults.audioMode
        audioDevice = ProcessInfo.processInfo.environment["CLATTERDRIVE_LAUNCHER_AUDIO_DEVICE"] ?? defaults.audioDevice
        status = "Stopped"
        isRunning = false
        logs = []
        driveProfileNames = ProfileCatalog.fallback.driveProfiles.map(\.name)
        acousticProfileNames = ProfileCatalog.fallback.acousticProfiles.map(\.name)

        backend.onLog = { [weak self] line in
            Task { @MainActor in
                self?.appendLog(line)
            }
        }
        backend.onReady = { [weak self] in
            Task { @MainActor in
                self?.isRunning = true
                self?.status = "Running"
            }
        }
        backend.onExit = { [weak self] _ in
            Task { @MainActor in
                self?.isRunning = false
                self?.status = "Stopped"
            }
        }
    }

    public var currentPort: Int {
        Int(portText) ?? 0
    }

    public var currentSettings: BackendSettings {
        BackendSettings(
            host: host,
            port: currentPort,
            backingDirectory: backingDirectory,
            audioMode: audioMode,
            audioDevice: audioDevice,
            driveProfile: driveProfile,
            acousticProfile: acousticProfile
        )
    }

    public var validationMessage: String {
        currentSettings.validationMessage() ?? ""
    }

    public var canStart: Bool {
        !isRunning && validationMessage.isEmpty
    }

    public var webDavURL: String {
        MountCommandBuilder.webDavURL(settings: currentSettings)
    }

    public var mountCommand: String {
        MountCommandBuilder.mountCommand(settings: currentSettings)
    }

    public var unmountCommand: String {
        MountCommandBuilder.unmountCommand()
    }

    public func refreshProfiles() {
        do {
            let catalog = try backend.loadProfiles()
            driveProfileNames = catalog.driveProfiles.map(\.name)
            acousticProfileNames = catalog.acousticProfiles.map(\.name)
            if !driveProfileNames.contains(driveProfile), let first = driveProfileNames.first {
                driveProfile = first
            }
            if !acousticProfileNames.contains(acousticProfile), let first = acousticProfileNames.first {
                acousticProfile = first
            }
        } catch {
            appendLog("Profile load failed: \(error.localizedDescription)")
        }
    }

    public func start() {
        let settings = currentSettings
        if let validation = settings.validationMessage() {
            status = validation
            return
        }
        do {
            try FileManager.default.createDirectory(
                atPath: settings.backingDirectory,
                withIntermediateDirectories: true
            )
            try backend.start(settings: settings)
            isRunning = true
            status = "Starting"
        } catch {
            status = "Start failed"
            appendLog(error.localizedDescription)
        }
    }

    public func stop() {
        backend.stop()
        isRunning = false
        status = "Stopped"
    }

    private func appendLog(_ line: String) {
        logs.append(line)
        if logs.count > 300 {
            logs.removeFirst(logs.count - 300)
        }
    }
}
